File: src/simulation/GameSimulation.cpp
#include "../lib/Date.cpp"
#include "./SimulationTree.cpp"
#include "./SimulationTask.cpp"

/*
	Shared simulation data will be stored in this struct.
	
	One major optimization that is integral to this simulation is that
	cells are only updated when one of their input-values have changed.
	This is achieved by giving each cell a list of indices of cells it outputs to,
	so they can be added to a list of cells to update during next
	simulation-step when an input-value changes.
*/
struct GameSimulation {
	/* The simulation tree created by expanding root block template. */
	SimulationTree simtree;
	/* List of task objects. */
	Vector<SimulationTask> tasks;
	/* True if tree should rebuild. */
	bool shouldRebuild;
	/* True if tree should be reset. (i.e. rebuild, but without saving values - NOT YET IMPLEMENTED) */
	bool shouldReset;
	/* True if simulation should update. */
	bool isRunning;
	/* Amount of accumulated simulation steps, based on elapsed time and simulation-speed. */
	u64 prev_time_ms;
	u64 curr_time_ms;
	u32 accumulated_msteps;
	
	GameSimulation() {
		this->shouldRebuild	= true;
		this->shouldReset	= true;
		this->isRunning		= true;
		this->prev_time_ms			= 0;
		this->curr_time_ms			= 0;
		this->accumulated_msteps	= 0;
	}
	
	// ============================================================
	// Helpers.
	// ------------------------------------------------------------
	
	void pushCellUpdate(u32 ind, u32 tgt, u32 val) {
		this->tasks[ind / CELLS_PER_TASK].pushCellUpdate(ind, tgt, val);
	}
	void pushCellUpdate(SimulationUpdate& upd) {
		this->tasks[upd.get_ind() / CELLS_PER_TASK].pushCellUpdate(upd);
	}

	u32 getCellValue(u32 ind, u32 tgt) {
		const SimulationTask& task = this->tasks[ind / CELLS_PER_TASK];
		return task.cell_buffer[task.toLocalIndex(ind)].values[tgt];
	}
	
	// ============================================================
	// Initialization stages.
	// ------------------------------------------------------------
	
	void generate_cell_data(
		SimulationTree& simtree,
		Vector<SimulationCell>& cell_buffer
	) {
		for(SimulationBlock& simblock : simtree.simblocks) {
			const BlockTemplate& btmp = simtree.getTemplate(simblock);
			for(const Cell& cell : btmp.cells) {
				const int index = cell_buffer.size();
				simblock.cmap[cell.id] = index;
				cell_buffer.push_back(SimulationCell(cell.value, cell.taskOrder));
			}
		}
	}
	
	void generate_link_data(
		SimulationTree& simtree,
		Vector<SimulationCell>& cell_buffer,
		Vector<SimulationLink>& link_buffer
	) {
		for(SimulationBlock& simblock : simtree.simblocks) {
			const BlockTemplate& btmp = simtree.getTemplate(simblock);
			Map<ItemId, Vector<SimulationLink>> gathered_links;
			// get links in this block which output from a cell in this block.
			{
				const auto& map = simtree.library.getOutputtingLinks(simblock.templateId)[ItemId::THIS_BLOCK];
				printf("<> gathered_links - map 1: %lu\n", map.size());
				for(const auto& [cellId, list] : map) {
					for(const Link& link : list) {
						const u32 ind = simtree.getCellIndex(simblock, link.dst.bid, link.dst.cid);
						const u32 tgt = link.dst.tgt;
						gathered_links[cellId].emplace_back(ind, tgt);
					}
				}
			}
			// get links in parent which output from a cell in this (child) block.
			printf("<> gathered_links - csimb: %llu\n", simblock.blockId);
			printf("<> gathered_links - ctemp: %llu\n", simblock.templateId);
			if(simblock.parentIndex != SimulationBlock::INDEX_NONE) {
				const SimulationBlock& parentSB = simtree.getParent(simblock);
				const auto& map = simtree.library.getOutputtingLinks(parentSB.templateId)[simblock.blockId];
				auto& supermap = simtree.library.getOutputtingLinks(parentSB.templateId);
				printf("<> gathered_links - map 2: %lu\n", map.size());
				for(const auto& [cellId, list] : map) {
					for(const Link& link : list) {
						printf("<> gathered_links - CID: %llu\n", cellId);
						const u32 ind = simtree.getCellIndex(parentSB, link.dst.bid, link.dst.cid);
						const u32 tgt = link.dst.tgt;
						printf("<> gathered_links - PRE: %llu\n", cellId);
						gathered_links[cellId].emplace_back(ind, tgt);
						printf("<> gathered_links - POST: %llu\n", cellId);
					}
				}
			}
			// add links to buffer and notify simcells.
			printf("<> gathered_links: %lu\n", gathered_links.size());
			for(const auto& [cellId, list] : gathered_links) {
				auto& cell = cell_buffer[simblock.cmap[cellId]];
				cell.links_len = list.size();
				cell.links_ofs = link_buffer.size();
				for(const SimulationLink simlink : list) link_buffer.push_back(simlink);
			}
		}
	}
	
	void generate_tasks(
		const Vector<SimulationCell>& cellbuf,
		const Vector<SimulationLink>& linkbuf
	) {
		this->tasks.clear();
		const u32 N = cellbuf.size();
		for(u32 n=0;n<N;n+=CELLS_PER_TASK) {
			// create task.
			const u32 ibeg = n;
			const u32 iend = std::min(N, ibeg + CELLS_PER_TASK);
			SimulationTask task = SimulationTask(cellbuf, linkbuf, ibeg, iend);
			// initialize and propagate initial cell values.
			for(int i=ibeg;i<iend;i++) task.initializeCellValue(i, cellbuf[i].values[LINK_TARGETS.OUTPUT]);
			tasks.push_back(task);
		}
	}
	
	/* Initialize or rebuild simulation data. */
	void rebuild(BlockTemplateLibrary& library, ItemId rootTemplateId) {
		printf("checkpoint 0\n");
		SimulationTree& oldTree = this->simtree;
		SimulationTree  newTree = SimulationTree(library, rootTemplateId);

		printf("checkpoint 1\n");
		Vector<SimulationCell> cell_buffer;
		cell_buffer.reserve(library.totalCellsInTree(rootTemplateId));
		this->generate_cell_data(newTree, cell_buffer);

		printf("checkpoint 2\n");
		Vector<SimulationLink> link_buffer;
		link_buffer.reserve(library.totalLinksInTree(rootTemplateId));
		this->generate_link_data(newTree, cell_buffer, link_buffer);

		printf("checkpoint 3\n");
		this->generate_tasks(cell_buffer, link_buffer);

		printf("checkpoint 4\n");
		this->simtree = newTree;
		this->shouldRebuild = false;
		this->shouldReset = false;
	}
	
	// ============================================================
	// Update
	// ------------------------------------------------------------
	void update(float simulationRate) {
		const u64 max_runtime_ms = 1000 / 60;
		u64 prev = this->prev_time_ms = this->curr_time_ms;
		u64 curr = this->curr_time_ms = Date::now_ms();
		if((prev == 0) | !this->isRunning) prev = curr;
		this->accumulated_msteps += (curr - prev) * simulationRate;
		if(this->isRunning) {
			while(this->accumulated_msteps > 1000) {
				// check if time limit reached
				//printf("update step 0\n");
				if(Date::now_ms() - curr < max_runtime_ms) {
					this->accumulated_msteps -= 1000;
				} else {
					this->accumulated_msteps = 0;
					break;
				}
				// perform simulation step.
				//printf("update step 1\n");
				for(SimulationTask& task : this->tasks) {
					if(task.shouldUpdate()) task.update();
				}
				// gather and spread outputs.
				//printf("update step 2\n");
				for(SimulationTask& task : this->tasks) {
					for(SimulationUpdate& upd : task.out_buffer) this->pushCellUpdate(upd);
				}
			}
		}
	}
};

File: src/simulation/GameSimulation.js

/*
	Shared simulation data will be stored in this struct.
	
	One major optimization that is integral to this simulation is that
	cells are only updated when one of their input-values have changed.
	This is achieved by giving each cell a list of indices of cells it outputs to,
	so they can be added to a list of cells to update during next
	simulation-step should an input-value change.
	
	------------------------------------------------------------
	Note that the performance, even despite considerable effort,
	is still very poor due to the lack of a type-system in JS.
	
	High performance will likely involve creating a local C++ server
	and sending rendering data between browser and server.
*/
class GameSimulation {
	// ============================================================
	// Structors.
	// ------------------------------------------------------------
	
	constructor() {
		/* Root simulation block of sim-block tree. */
		this.root_simulation_block	= null;
		/* Array of task objects. */
		this.tasks					= null;
		/* True if tree should rebuild. */
		this.shouldRebuild			= true;
		/* True if tree should be reset. (rebuild, but without saving values - NOT YET IMPLEMENTED) */
		this.shouldReset			= true;
		/* True if simulation should update. */
		this.isRunning				= true;
		/* Amount of accumulated simulation steps, based on elapsed time and simulation-speed. */
		this.accumulatedSteps		= 0;
		this.previousTime			= 0;
	}
	
	// ============================================================
	// Helpers.
	// ------------------------------------------------------------
	
	getTaskIndex(cellIndex) {
		return Math.floor(cellIndex / CELLS_PER_TASK);
	}
	
	pushCellUpdate(ind, tgt, val) {
		this.tasks[this.getTaskIndex(ind)].pushCellUpdate(ind, tgt, val);
	}
	
	applyCellOutputChange(ind, val) {
		this.tasks[this.getTaskIndex(ind)].applyCellOutputChange(ind, val);
	}
	
	// ============================================================
	// Initialization stages.
	// ------------------------------------------------------------
	
	/* Generate cell data. */
	rebuild_generateCellData(simblock, cell_buffer) {
		const template = simblock.template;
		for(const cell of template.cells) {
			const cind = simblock.getCellIndex(cell.id);
			const typeInd = TASK_CELL_TYPE_ORDER_MAP.get(cell.type);
			cell_buffer.set_all(cind, cell.value, 0x0, 0x0, typeInd, 0, 0);
		}
		for(const block of template.blocks) {
			const sb = simblock.getSimblock(block.id);
			this.rebuild_generateCellData(sb, cell_buffer);
		}
	}
	
	/*
		Recursively generate link data.
		
		NOTE: this is generated after simblock tree and cells
		because it sometimes requires cells in siblings during linkgen.
	*/
	rebuild_generateSimblockLinks(simblock, parentSB, cell_buffer, link_buffer) {
		this.rebuild_generateLinkData(simblock, parentSB, cell_buffer, link_buffer);
		for(const block of simblock.template.blocks) {
			const sb = simblock.getSimblock(block.id);
			this.rebuild_generateSimblockLinks(sb, simblock, cell_buffer, link_buffer);
		}
	}
	
	/*
		Generate link data for links that output from given simblock.
		NOTE: this function sometimes requires cell-data from simblock-siblings.
	*/
	rebuild_generateLinkData(simblock, parentSB, cell_buffer, link_buffer) {
		// get links from self which have a cell in this block as an output.
		const s_data = simblock.template.getOutputtingLinks(ComponentId.THIS_BLOCK);
		// get links from parent which have a cell in this block as an output.
		const p_data = parentSB ? parentSB.template.getOutputtingLinks(simblock.blockId) : null;
		// add links to buffer and set cell link-lists.
		for(const cell of simblock.template.cells) {
			const cid = cell.id;
			const cind = simblock.getCellIndex(cid);
			const s_arr = s_data?.get(cid);
			const p_arr = p_data?.get(cid);
			let len = 0;
			let ofs = link_buffer.count;
			if(s_arr) for(const [bid_dst, cid_dst, tgt_dst] of s_arr) {
				const ind_dst = simblock.getSimblock(bid_dst).getCellIndex(cid_dst);
				link_buffer.set_all(ofs+len, ind_dst, tgt_dst);
				len++;
			}
			if(p_arr) for(const [bid_dst, cid_dst, tgt_dst] of p_arr) {
				const ind_dst = parentSB.getSimblock(bid_dst).getCellIndex(cid_dst);
				link_buffer.set_all(ofs+len, ind_dst, tgt_dst);
				len++;
			}
			cell_buffer.setLinkList(cind, len, ofs);
			link_buffer.count += len;
		}
	}
	
	/* create new array of task objects. */
	rebuild_createTasks(cellbuf, linkbuf) {
		const tasks = [];
		const N = cellbuf.count;
		for(let ibeg=0; ibeg<N; ibeg+=CELLS_PER_TASK) {
			let iend = Math.min(N, ibeg + CELLS_PER_TASK);
			// get total number of links.
			let link_count = 0;
			for(let i=ibeg;i<iend;i++) link_count += cellbuf.get_links_len(i);
			// create task.
			const task = new SimulationTask(ibeg, iend, link_count);
			tasks.push(task);
			// copy cells to task-local buffer.
			for(let i=ibeg;i<iend;i++) task.cell_buffer.copyFrom(cellbuf, i-ibeg, i);
			// copy links to task-local buffer, and update link-list pointers.
			let newofs = 0;
			for(let i=ibeg;i<iend;i++) {
				const len = cellbuf.get_links_len(i);
				const ofs = cellbuf.get_links_ofs(i);
				task.cell_buffer.set_links_ofs(i-ibeg, newofs);
				for(let x=0;x<len;x++) task.link_buffer.copyFrom(linkbuf, newofs++, ofs+x);
			}
			task.link_buffer.count = newofs;
			// initialize and propagate initial cell values.
			for(let i=ibeg;i<iend;i++) task.initializeCellValue(i, cellbuf.get_value_out(i));
		}
		this.tasks = tasks;
	}
	
	/* Initialize or rebuild simulation data. */
	rebuild() {
		const t0 = Date.now();
		// create buffers.
		const rootTemplate = gameData.rootBlock.template;
		const cell_buffer = new SimulationCellBuffer(rootTemplate.totalCellsInTree());
		const link_buffer = new SimulationLinkBuffer(rootTemplate.totalLinksInTree());
		// build simulation data, as well as new sim-block tree, starting from bottom of tree and finishing with root block.
		const oldTree = this.root_simulation_block;
		const newTree = new SimulationTreeBlock(gameData.rootBlock.template, ComponentId.NONE, 0);
		this.root_simulation_block = newTree;
		this.rebuild_generateCellData(newTree, cell_buffer);
		const numCells = cell_buffer.count = newTree.numCells;
		console.log("numCells", numCells);
		// build simulation links once tree and cell data have been generated.
		this.rebuild_generateSimblockLinks(newTree, null, cell_buffer, link_buffer);
		const numLinks = link_buffer.count;
		console.log("numLinks", numLinks);
		// create new task objects.
		this.rebuild_createTasks(cell_buffer, link_buffer);
		console.log("rebuild_createTasks", this.tasks.length);
		// finishing touches.
		this.shouldRebuild = false;
		this.shouldReset = false;
		const t1 = Date.now();
		console.log("simulation.rebuild", t1-t0, numCells, numLinks);
	}
	
	// ============================================================
	// Update stages.
	// ------------------------------------------------------------
	
	update_gatherAndSpreadBuffers() {
		// gather and spread updates.
		for(const task of this.tasks) {
			const N = task.out_buffer.count;
			for(let x=0;x<N;x++) {
				const ind = task.out_buffer.get_ind(x);
				const tgt = task.out_buffer.get_tgt(x);
				const val = task.out_buffer.get_val(x);
				this.pushCellUpdate(ind, tgt, val);
			}
		}
	}
	
	update_tasks() {
		// TODO: make multi-threaded.
		for(const task of this.tasks) task.update();
	}
	
	update() {
		if(this.previousTime === 0 | !this.isRunning) this.previousTime = Date.now();
		const prev = this.previousTime;
		const curr = Date.now();
		const runlimit = 1000 / 60;// safety: minimum fps target.
		this.accumulatedSteps += (curr - prev) * gameData.simulationSpeed / 1000;
		this.previousTime = curr;
		
		if(this.isRunning) {
			while(this.accumulatedSteps > 1) {
				// check if time limit reached
				if(Date.now() - curr < runlimit) { this.accumulatedSteps -= 1.0; }
				else { this.accumulatedSteps = 0.0; break; }
				// perform simulation step.
				const t0 = Date.now();
				this.update_tasks();
				const t1 = Date.now();
				this.update_gatherAndSpreadBuffers();
				const t2 = Date.now();
				Performance.increment_time("simulation.update_tasks  ", t1-t0);
				Performance.increment_time("simulation.update_gth+spr", t2-t1);
				for(const task of this.tasks) {
					Performance.increment_count("simulation.task.perf_n_updates", task.perf_n_updates); task.perf_n_updates=0;
					Performance.increment_count("simulation.task.perf_n_outputs", task.perf_n_outputs); task.perf_n_outputs=0;
				}
				Performance.increment_time("simulation.update        ", t2-t0);
			}
		}
	}
	
	// ============================================================
	// Content change handlers.
	// Note: It is assumed that content is being added to root-block.
	// ------------------------------------------------------------
	onContentChanged_addCell(cell) {
		this.shouldRebuild = true;
		// TODO: add cell to task and rebuild task only.
	}
	onContentChanged_remCell(cell) {
		this.shouldRebuild = true;
		// TODO: remove cell to task and rebuild task only.
	}
	onContentChanged_addLink(link) {
		this.shouldRebuild = true;
		// TODO: add link to task and rebuild task only.
	}
	onContentChanged_remLink(link) {
		this.shouldRebuild = true;
		// TODO: remove link from task and rebuild task only.
	}
	onContentChanged_addBlock() { this.shouldRebuild = true; }
	onContentChanged_remBlock() { this.shouldRebuild = true; }
	
};





